addons/mrp/mrp.py

```diff
@@ -805,12 +805,22 @@ def _compute_costs_from_production(self, cr, uid, ids, context=None):
         """ Generate workcenter costs and rectify the value of the quant
 
         Must be called at the end of the production.
-        The value of the quant was not possible to compute before the end of
-        the manufacturing order due to the cost of raw manterial and production
-        costs. The price of the quant was not set in get_price_unit.
+        The full value of the quant was not possible to compute before the end of
+        the manufacturing order due production costs not ended yet.
         """
         for production in self.browse(cr, uid, ids):
-            total_cost = self._costs_generate(cr, uid, production)
+            production_cost = self._costs_generate(cr, uid, production)
+            if not production_cost:
+                continue
+            average_cost = production_cost / production.product_qty
+            # may have been produced in multiple steps (multiple quants)
+            for produced_product in production.move_created_ids2:
+                # only the produced product
+                if produced_product.product_id == production.product_id:
+                    # should only have one quant at the end of the production
+                    increased_value = produced_product.quant_ids.inventory_value + average_cost * produced_product.quant_ids.qty
+                    increased_cost = increased_value / produced_product.quant_ids.qty
+                    self.pool['stock.quant'].write(cr, SUPERUSER_ID, produced_product.quant_ids.ids, {'cost': increased_cost}, context=context)
```

Comment:
I try to avoid writing on quants to support the real-time valuation, so why do you do it here?  And I don't see why you should not be able to calculate the workcenter costs before.  I can not agree with this...